### DISCUSSION TITLE
Is mounted() called after a cached payload (full static mode) is returned?

### DISCUSSION THREAD

**Discussion by hecktarzuli** [nuxt/nuxt#7951] [General]
I'm trying to integrate speedcurve into a full static site and I'm curious: 

a) is mounted() called after the cached payload is fetched and loaded by the browser?
or
b) does it act just like generate & SSR where the page is mounted() and fetch ends whenever it is done (async)

If b), then I don't know of any real way we could fairly stop a performance timer as there wouldn't be a way for us to know when the page is finished loading when using full static mode.

@pi0

**Comment by pi0** [comment]
**For SSR:**

On initial page load, `mounted()` (as well as other life-cycle hooks of page component) are always called **after** payload (aka cache) is downloaded and loaded.

  - If you check the `<head>` section, we do preload the chunk for cached payload 
  - As soon as main app is being created, the chunk will be executed and loaded into memory ([see here](https://github.com/nuxt/nuxt.js/blob/5e9d759d7bee6a3d32e16ef05eef988e882c600a/packages/vue-app/template/client.js#L833))

We need an inline script to more precisely measure `FCP` (and even `LCP` if nothing is lazy loaded) and for FID, we can also use `window.onNuxtReady` as this hook is called after main app is `mounted` and [web-vitals](https://github.com/GoogleChrome/web-vitals) package to leverage browser performance API.

Example: (`plugins/performance.client.js`):

```js
import { getFCP, getFID } from 'web-vitals'

window.onNuxtReady(($nuxt) => {
  // Nuxt App Mounted so we can be almost sure measurements here are final
  getFCP(({ value }) => { console.log('First Contentful paint:', value) })
  getFID(({ value }) => { console.log('First Input delay:', value) })
})
```

**For CSR:** (Navigating on client-side with `<NuxtLink>`)

- When using server mode, `fetch()` is also called before `mounted()` but running in parallel and not blocking navigation. It highly depends on how one use `fetch` it may or may not delay LCP (like if having a big loading overlay), FID (like if forms pending for data to be reactive) and also affect CLS (if placeholder for `$fetchState` is not sized properly)
- When using full-static target, we **behave fetch similar to asyncData** and block next page load until payload is loaded ([src](https://github.com/nuxt/nuxt.js/blob/5e9d759d7bee6a3d32e16ef05eef988e882c600a/packages/vue-app/template/client.js#L509)) this means also for CSR+FullStatic mounted is called after payload is being called. But it is usually fast because next chunks are also prefetched in background.

-> Generally because of this complexities probably it does not worth to report CSR data.

**Reply by hecktarzuli** [reply]
For CSR tracking, this is good news.  This means that with full static, users can safely put the 'stop beacon' in mounted() and know the page is rendered (probably with $nextTick internally)

For universal with asyncData, they can put the 'stop beacon' in mounted()

For universal or classic 'generate' with fetch, they can put the 'stop beacon' in fetch()

I was hoping to do some of these things automagically, but I think there are too many ways it would break.. the only 'safe' one seems to be full static.  I'll think some more, thanks!

**Reply by pi0** [reply]
I'm not sure meant any of above conclusions but do as you wish :)
